File: backend/app/core/security.py
from datetime import datetime, timedelta
from typing import Optional, List
import bcrypt
from jose import JWTError, jwt
from fastapi import Depends, HTTPException, status
from fastapi.security import OAuth2PasswordBearer
from sqlalchemy.orm import Session
from sqlalchemy import or_
from app.core.database import get_db

SECRET_KEY = "simnexus-secret-key-change-in-production"
ALGORITHM = "HS256"
ACCESS_TOKEN_EXPIRE_MINUTES = 60 * 24  # 24 hours

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="/api/auth/login")


def hash_password(password: str) -> str:
    return bcrypt.hashpw(password.encode(), bcrypt.gensalt()).decode()


def verify_password(plain: str, hashed: str) -> bool:
    return bcrypt.checkpw(plain.encode(), hashed.encode())


def create_access_token(data: dict, expires_delta: Optional[timedelta] = None) -> str:
    to_encode = data.copy()
    expire = datetime.utcnow() + (expires_delta or timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES))
    to_encode["exp"] = expire
    return jwt.encode(to_encode, SECRET_KEY, algorithm=ALGORITHM)


def get_current_user(token: str = Depends(oauth2_scheme), db: Session = Depends(get_db)):
    from app.models.user import User
    credentials_exc = HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="无效的认证凭证",
        headers={"WWW-Authenticate": "Bearer"},
    )
    try:
        payload = jwt.decode(token, SECRET_KEY, algorithms=[ALGORITHM])
        username: str = payload.get("sub")
        if username is None:
            raise credentials_exc
    except JWTError:
        raise credentials_exc

    user = db.query(User).filter(User.username == username, User.is_active == True).first()
    if user is None:
        raise credentials_exc
    return user


def require_admin(current_user=Depends(get_current_user)):
    from app.models.user import UserRole
    if current_user.role != UserRole.ADMIN:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="需要管理员权限")
    return current_user


def _perm(user):
    """Return merged permission dict from user's RBAC roles.
    Admin users bypass all checks — callers should check user.role == ADMIN first.
    Returns None if user has no roles assigned.
    """
    from app.models.user import UserRole
    if user.role == UserRole.ADMIN:
        return {
            "can_view_sim": True,
            "can_approve_requests": True,
            "can_view_history": True,
            "can_support": True,
            "read_only": False,
            "allowed_modem_ids": None,
        }
    roles = getattr(user, "rbac_roles", None) or []
    if not roles:
        return None

    if any(r.allowed_modem_ids is None for r in roles):
        modem_ids = None
    else:
        modem_ids = list({mid for r in roles for mid in (r.allowed_modem_ids or [])}) or None

    return {
        "can_view_sim":         any(r.can_view_sim for r in roles),
        "can_approve_requests": any(r.can_approve_requests for r in roles),
        "can_view_history":     any(r.can_view_history for r in roles),
        "can_support":          any(r.can_support for r in roles),
        "read_only":            all(r.read_only for r in roles),
        "allowed_modem_ids":    modem_ids,
    }


def get_user_modem_grants(user_id: int, db: Session, level: Optional[str] = None, user=None) -> List[int]:
    """Return modem IDs the user has access to.
    Queries sim_grants (current effective grants) plus role-based auto-grants.
    level=None → any grant (view or use)
    level='use' → only use-level grants
    """
    from app.models.sim_request import SimGrant, PermissionLevel
    from app.models.modem import Modem
    now = datetime.utcnow()
    q = db.query(SimGrant.modem_id).filter(
        SimGrant.user_id == user_id,
        or_(SimGrant.expires_at.is_(None), SimGrant.expires_at > now),
    )
    if level == "use":
        q = q.filter(SimGrant.granted_level == PermissionLevel.USE)
    grant_ids = set(r.modem_id for r in q.all())

    if user is not None:
        roles = getattr(user, "rbac_roles", None) or []
        for role in roles:
            if role.can_approve_requests:
                # Approvers automatically have use-level access to their managed cards
                if role.allowed_modem_ids is None:
                    all_ids = [r.id for r in db.query(Modem.id).all()]
                    grant_ids.update(all_ids)
                else:
                    grant_ids.update(role.allowed_modem_ids)
            elif role.allowed_modem_ids is not None:
                # Non-approver roles: explicit allowed_modem_ids = auto-grant access to those cards
                grant_ids.update(role.allowed_modem_ids)

    return list(grant_ids)


def require_approve_requests(current_user=Depends(get_current_user)):
    from app.models.user import UserRole
    if current_user.role == UserRole.ADMIN:
        return current_user
    p = _perm(current_user)
    if not p or not p.get("can_approve_requests"):
        raise HTTPException(status_code=403, detail="无审批权限")
    return current_user


def require_view_history(current_user=Depends(get_current_user)):
    from app.models.user import UserRole
    if current_user.role == UserRole.ADMIN:
        return current_user
    p = _perm(current_user)
    if not p or not p.get("can_view_history"):
        raise HTTPException(status_code=403, detail="无短信记录查看权限")
    return current_user


def is_support_staff(user) -> bool:
    from app.models.user import UserRole
    if user.role == UserRole.ADMIN:
        return True
    roles = getattr(user, "rbac_roles", None)
    if roles and any(r.can_support for r in roles):
        return True
    return False


def require_support_staff(current_user=Depends(get_current_user)):
    if not is_support_staff(current_user):
        raise HTTPException(status_code=403, detail="无客服权限")
    return current_user

File: backend-go/security/security.go
package security

import (
	"time"

	"simnexus-go/config"
	"simnexus-go/models"

	"github.com/golang-jwt/jwt/v5"
	"golang.org/x/crypto/bcrypt"
	"gorm.io/gorm"
)

// accessTokenExpireMinutes JWT 令牌有效期（24 小时）。
const accessTokenExpireMinutes = 60 * 24 // 24h

// HashPassword 使用 bcrypt 对密码进行散列，返回可直接存储的字符串。
func HashPassword(pw string) (string, error) {
	b, err := bcrypt.GenerateFromPassword([]byte(pw), bcrypt.DefaultCost)
	return string(b), err
}

// VerifyPassword checks a plaintext password against a bcrypt hash.
func VerifyPassword(plain, hashed string) bool {
	return bcrypt.CompareHashAndPassword([]byte(hashed), []byte(plain)) == nil
}

// CreateAccessToken issues a JWT with `sub` = username.
func CreateAccessToken(username string) (string, error) {
	claims := jwt.MapClaims{
		"sub": username,
		"exp": time.Now().Add(accessTokenExpireMinutes * time.Minute).Unix(),
	}
	tok := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return tok.SignedString([]byte(config.SecretKey))
}

// ParseToken validates a JWT and returns the `sub` claim (username).
func ParseToken(tokenStr string) (string, error) {
	tok, err := jwt.Parse(tokenStr, func(t *jwt.Token) (interface{}, error) {
		return []byte(config.SecretKey), nil
	})
	if err != nil || !tok.Valid {
		if err == nil {
			err = jwt.ErrTokenInvalidClaims
		}
		return "", err
	}
	claims, ok := tok.Claims.(jwt.MapClaims)
	if !ok {
		return "", jwt.ErrTokenInvalidClaims
	}
	sub, _ := claims["sub"].(string)
	if sub == "" {
		return "", jwt.ErrTokenInvalidClaims
	}
	return sub, nil
}

// LoadUserByUsername loads an active user with RBAC roles (and their modem scope) joined.
func LoadUserByUsername(db *gorm.DB, username string) (*models.User, error) {
	var u models.User
	err := db.Preload("RbacRoles").Preload("RbacRoles.ModemScope").
		Where("username = ? AND is_active = ?", username, true).First(&u).Error
	if err != nil {
		return nil, err
	}
	return &u, nil
}

// Permissions 是用户合并后的有效权限集，镜像 Python 的 _perm() 返回值。
// AllowedModemIDs=nil 且 unrestricted=true 表示不限设备范围。
type Permissions struct {
	CanViewSim         bool
	CanApproveRequests bool
	CanViewHistory     bool
	CanSupport         bool
	ReadOnly           bool
	// AllowedModemIDs: nil 表示无限制；非 nil 为受限设备 ID 集合。
	AllowedModemIDs []uint
	unrestricted    bool // 内部标记：true 时 AllowedModemIDs 语义上为 nil（无限制）
}

// Perm 计算用户的合并权限。管理员返回全权限；无 RBAC 角色的用户返回 nil。
// 正向权限标志取所有角色的并集（any），ReadOnly 取交集（all）。
func Perm(u *models.User) *Permissions {
	if u.IsAdmin() {
		return &Permissions{
			CanViewSim: true, CanApproveRequests: true, CanViewHistory: true,
			CanSupport: true, ReadOnly: false, unrestricted: true,
		}
	}
	roles := u.RbacRoles
	if len(roles) == 0 {
		return nil
	}

	p := &Permissions{ReadOnly: true}
	// positive flags = any; read_only = all
	for _, r := range roles {
		p.CanViewSim = p.CanViewSim || r.CanViewSim
		p.CanApproveRequests = p.CanApproveRequests || r.CanApproveRequests
		p.CanViewHistory = p.CanViewHistory || r.CanViewHistory
		p.CanSupport = p.CanSupport || r.CanSupport
		if !r.ReadOnly {
			p.ReadOnly = false
		}
	}

	// modem scope: only approver roles matter
	var approverRoles []models.Role
	for _, r := range roles {
		if r.CanApproveRequests {
			approverRoles = append(approverRoles, r)
		}
	}
	if len(approverRoles) > 0 {
		anyEmpty := false
		set := map[uint]struct{}{}
		for _, r := range approverRoles {
			if len(r.ModemScope) == 0 {
				anyEmpty = true
			}
			for _, m := range r.ModemScope {
				set[m.ID] = struct{}{}
			}
		}
		if anyEmpty || len(set) == 0 {
			p.unrestricted = true
		} else {
			for id := range set {
				p.AllowedModemIDs = append(p.AllowedModemIDs, id)
			}
		}
	} else {
		p.unrestricted = true
	}
	return p
}

// IsUnrestrictedScope reports whether the permission scope is unrestricted (None in Python).
func (p *Permissions) IsUnrestrictedScope() bool { return p == nil || p.unrestricted }

// IsSupportStaff reports whether the user is admin or has a can_support role.
func IsSupportStaff(u *models.User) bool {
	if u.IsAdmin() {
		return true
	}
	for _, r := range u.RbacRoles {
		if r.CanSupport {
			return true
		}
	}
	return false
}

// GetUserModemGrants 返回用户可访问的设备 ID 集合。
// level="" 表示任意权限级别，level="use" 仅返回可发送短信的设备。
// 审批员角色在其管辖范围内自动获得 use 级别授权（无需申请）。
func GetUserModemGrants(db *gorm.DB, userID uint, level string, u *models.User) []uint {
	now := time.Now().UTC()
	q := db.Model(&models.SimGrant{}).
		Where("user_id = ? AND (expires_at IS NULL OR expires_at > ?)", userID, now)
	if level == models.LevelUse {
		q = q.Where("granted_level = ?", models.LevelUse)
	}
	var grants []models.SimGrant
	q.Find(&grants)

	set := map[uint]struct{}{}
	for _, g := range grants {
		set[g.ModemID] = struct{}{}
	}

	if u != nil {
		for _, role := range u.RbacRoles {
			scopeIDs := make([]uint, 0, len(role.ModemScope))
			for _, m := range role.ModemScope {
				scopeIDs = append(scopeIDs, m.ID)
			}
			if role.CanApproveRequests {
				if len(scopeIDs) == 0 {
					// unrestricted approver → all modems
					var all []models.Modem
					db.Select("id").Find(&all)
					for _, m := range all {
						set[m.ID] = struct{}{}
					}
				} else {
					for _, id := range scopeIDs {
						set[id] = struct{}{}
					}
				}
			} else if len(scopeIDs) > 0 {
				for _, id := range scopeIDs {
					set[id] = struct{}{}
				}
			}
		}
	}

	out := make([]uint, 0, len(set))
	for id := range set {
		out = append(out, id)
	}
	return out
}

// ContainsUint 判断 uint 切片中是否包含指定值。
func ContainsUint(s []uint, v uint) bool {
	for _, x := range s {
		if x == v {
			return true
		}
	}
	return false
}
